### PR TITLE
net: tracing: Add socket tracing support

### PR DESCRIPTION
**Pull request by jukkar**
If network socket tracing is enabled, then the system will track various socket API calls for usage.

### CODE REVIEW

**Comment by nashif** [comment]
@jukkar  nice, while playing with this, discovered some missing hooks and added them here: #73674, not sure if you hit the same issue.

**Comment by jukkar** [comment]
> not sure if you hit the same issue.

No, I did not see any issue with those missing hooks. I will take your commit to this PR in order to avoid conflicts later.

**Comment by nashif** [comment]
> No, I did not see any issue with those missing hooks. I will take your commit to this PR in order to avoid conflicts later.

I got this with tests/net/socket/tcp/ and enabling tracing...

**Review by cfriedt — DISMISSED** [review]
🪁

**Comment by jukkar** [comment]
> I got this with tests/net/socket/tcp/ and enabling tracing...

Thanks, that issue is now fixed

File: include/zephyr/tracing/tracing.h [hunk 1]
```diff
@@ -2026,6 +2026,304 @@
...
+#define sys_port_trace_socket_listen_exit(socket, ret)
+
+/**
+ * @brief Trace accept of network sockets
+ * @param socket Socket object
+ */
+#define sys_port_trace_socket_accept_enter(socket)
+
+/**
+ * @brief Trace network socket listen attempt
```
Comment (rlubos) [inline include/zephyr/tracing/tracing.h:2122]: s/listen/accept

File: include/zephyr/tracing/tracing.h [hunk 2]
```diff
@@ -2026,6 +2026,304 @@
...
+ * @param socket Socket object
+ * @param max_len Maximum length of the data we can receive
+ * @param flags Flags for this receive operation
+ * @param addr Remote network address
+ * @param addrlen Network address length
+ */
+#define sys_port_trace_socket_recvfrom_enter(socket, max_len, flags, addr, addrlen)
+
+/**
+ * @brief Trace network socket sendmsg attempt
```
Comment (rlubos) [inline include/zephyr/tracing/tracing.h:2173]: s/sendmsg/recvfrom

File: include/zephyr/tracing/tracing.h [hunk 3]
```diff
@@ -2026,6 +2026,304 @@
...
+ * @param socket Socket object
+ * @param level Option level
+ * @param optname Option name
+ * @param optval Option value
+ * @param optlen Option value length
+ */
+#define sys_port_trace_socket_setsockopt_enter(socket, level, optname, optval, optlen)
+
+/**
+ * @brief Trace network socket getsockopt attempt
```
Comment (rlubos) [inline include/zephyr/tracing/tracing.h:2272]: s/getsockopt/setsockopt

**Review by rlubos — DISMISSED** [review]
LGTM, just a few nits in docs

**Review by rlubos — APPROVED** [review]

**Comment by jukkar** [comment]
I ran the tests/net twister suite manually with tracing enabled, and did not spot any issues any more.

**Comment by jukkar** [comment]
Twister fails with unrelated issues in the tree, upstream seems to be broken again
```
west build -p -b native_sim tests/subsys/sensing -T subsys.sensing -d ../build/sensing -t run
...
[113/115] Building native simulator runner, and linking final executable
FAILED: zephyr/CMakeFiles/native_runner_executable zephyr/zephyr.exe .../zephyrproject/build/sensing/zephyr/CMakeFiles/native_runner_executable .../zephyrproject/build/sensing/zephyr/zephyr.exe 
cd .../zephyrproject/build/sensing/zephyr && /usr/bin/make -f .../zephyrproject/zephyr/scripts/native_simulator/Makefile all --warn-undefined-variables -r NSI_CONFIG_FILE=.../zephyrproject/build/sensing/zephyr/NSI/nsi_config
/usr/bin/ld: .../zephyrproject/build/sensing/zephyr/NSI/.../zephyrproject/build/sensing/zephyr/zephyr.elf.loc_cpusw.o:.../zephyrproject/zephyr/drivers/sensor/bosch/bmi160/bmi160.c:1233: undefined reference to `__device_dts_ord_35'
/usr/bin/ld: .../zephyrproject/build/sensing/zephyr/NSI/.../zephyrproject/build/sensing/zephyr/zephyr.elf.loc_cpusw.o:.../zephyrproject/zephyr/drivers/sensor/bosch/bmi160/bmi160.c:1233: undefined reference to `__device_dts_ord_32'
collect2: error: ld returned 1 exit status
```

**Review by cfriedt — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by jukkar** [comment]
ping @nashif, are you ok with the changes?

**Review by nashif — APPROVED** [review]

**Pull request merged**